packages/eslint-plugin/src/rules/prefer-at.ts

```diff
@@ -0,0 +1,219 @@
+          if (node.property.type === AST_NODE_TYPES.Literal) {
+            return `${getFullName(node.object)}[${node.property.raw}]`;
+          }
+          return `${getFullName(node.object)}.${getFullName(node.property)}`;
+        default:
+          throw new UnknownNodeError(node);
+      }
+    }
+
+    function hasCallExpression(node: TSESTree.MemberExpression): boolean {
+      return (
+        node.object.type === AST_NODE_TYPES.CallExpression ||
+        (node.object.type === AST_NODE_TYPES.MemberExpression &&
+          hasCallExpression(node.object))
+      );
+    }
+
+    function getTypeAtLocation(node: TSESTree.Node): ts.Type | undefined {
+      const originalNode = parserServices.esTreeNodeToTSNodeMap.get(node);
+      if (!originalNode) {
+        return undefined;
+      }
+      return checker.getTypeAtLocation(originalNode);
+    }
+
+    type SupportedObject = (type: ts.Type) => boolean;
+
+    function checkObjectName(name: string): SupportedObject {
+      return type => type.getSymbol()?.name === name;
+    }
+
+    function checkObjectType(flags: ts.TypeFlags): SupportedObject {
+      return type => type.getFlags() === flags;
+    }
+
+    const supporterObjects: Array<SupportedObject> = [
+      checkObjectName('Array'),
+      checkObjectName('Int8Array'),
+      checkObjectName('Uint8Array'),
+      checkObjectName('Uint8ClampedArray'),
+      checkObjectName('Int16Array'),
+      checkObjectName('Uint16Array'),
+      checkObjectName('Int32Array'),
+      checkObjectName('Float32Array'),
+      checkObjectName('Uint32Array'),
+      checkObjectName('Float64Array'),
+      checkObjectName('BigInt64Array'),
+      checkObjectName('BigUint64Array'),
+      // eslint-disable-next-line @typescript-eslint/internal/prefer-ast-types-enum
+      checkObjectName('String'),
+      checkObjectType(ts.TypeFlags.String),
+    ];
+
+    function isSupportedObject(type: ts.Type): boolean {
+      return supporterObjects.some(check => check(type));
+    }
+
+    function isExpectedObject(
+      node: TSESTree.Node,
+    ): node is TSESTree.MemberExpression {
+      if (
+        node.type !== AST_NODE_TYPES.MemberExpression ||
+        (options.ignoreFunctions && hasCallExpression(node))
+      ) {
+        return false;
+      }
+      const type = getTypeAtLocation(node.object);
+      if (type && !isSupportedObject(type)) {
+        return false;
+      }
+      const atMember = type?.getProperty('at');
+      return (
+        atMember?.declarations?.every(declaration => {
+          const sourceFile = declaration.getSourceFile();
+          const directory = path.join(sourceFile.fileName, '../');
+          return directory.endsWith(path.join('node_modules/typescript/lib/'));
+        }) ?? false
+      );
+    }
+
+    function isExpectedExpressionLeft(
+      node: TSESTree.BinaryExpression,
+    ): boolean {
+      if (!isExpectedObject(node.left) || getName(node.left) !== 'length') {
```

Comment:
[Question] Similar lint rules to this one use the `getStaticValue` from `eslint-utils`. Have you tried using it in this file? I'm wondering if it'll help do some of the logical checks that are implemented manually right now.

Reply:
Bump for this unresolved thread @sviat9440 ?

Reply:
Behaviorally, it should help catch cases like this:

* `str[str['length'] - 1]`
* `const key = 'length'; str[str[key] - 1]`

For code, it would eliminate the need to manually write a `function getName`. Which is something a lot of our rules end up wanting. So we'd really rather than write explicit _"what is the string name of this property?"_ logic repeatedly without a real need.

https://eslint-community.github.io/eslint-utils/api/ast-utils.html#getstaticvalue

Reply:
(also, for anybody seeing just this thread and wondering why we're bumping at each other 😂 the comments were pending)